### PR TITLE
Introduced sign change in the preconditioners of ex22 and ex22p

### PR DESCRIPTION
**Pull request by psocratis**
Small change (sign change)  in the preconditioner that brings down the iteration numbers. 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1321](https://github.com/mfem/mfem/pull/1321) | @psocratis | @tzanio | @v-dobrev + @mlstowell | 02/25/20 | 02/27/20 | 02/29/20 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Can you elaborate on this? Was it changed before, why do we expect better convergence?

**Comment by psocratis** [comment]
Hi @Tzanio, 

The sign flip makes the Operator symmetric regardless of what convention was chosen by the user. 90% of the sample runs of the example converge must faster now, but I don't really have a concrete explanation, other than the fact that the Operator now has a real spectrum. This is the same change that it is introduced in #1279, but after discussion I had with @v-dobrev , we thought it would be better if the change can be included in the 4.1 release. 

Socratis

**Comment by tzanio** [comment]
> The sign flip makes the Operator symmetric regardless of what convention was chosen by the user. 90% of the sample runs of the example converge must faster now, but I don't really have a concrete explanation, other than the fact that the Operator now has a real spectrum.

Is this related to the result in Appendix A of https://library.seg.org/doi/pdf/10.1190/geo2015-0013.1?

**Comment by psocratis** [comment]
> > The sign flip makes the Operator symmetric regardless of what convention was chosen by the user. 90% of the sample runs of the example converge must faster now, but I don't really have a concrete explanation, other than the fact that the Operator now has a real spectrum.
> 
> Is this related to the result in Appendix A of https://library.seg.org/doi/pdf/10.1190/geo2015-0013.1?

I don't see a direct relation, since the block matrices in the example are not SPD in general. They are just symmetric.

**Comment by v-dobrev** [comment]
Comparing to `master` I saw two cases where the iterations increased a little:
1. In the sample run
```console
./ex22 -m ../data/inline-quad.mesh -o 3 -p 2 -no-vis
```
the iterations increase from 42 to 88.

2. In the sample run
```console
mpirun -np 4 ./ex22p -m ../data/inline-segment.mesh -o 3 -no-vis
```
the iterations increase from 39 to 78.

In all other cases the iterations are either the same or improved (significantly better in a number of cases).

**Comment by tzanio** [comment]
@mlstowell, can you take a quick look?

**Comment by tzanio** [comment]
> The source and/or boundary conditions don't seem to be applied correctly on the inline hex mesh. For example, see: `ex22 -m ../data/inline-hex.mesh -o 2 -p 1`.

Lets investigate and fix in a separate PR...

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by psocratis** [comment]
> > The source and/or boundary conditions don't seem to be applied correctly on the inline hex mesh. For example, see: `ex22 -m ../data/inline-hex.mesh -o 2 -p 1`.
> 
> Lets investigate and fix in a separate PR...

I'll take a look

**Comment by psocratis** [comment]
@tzanio , The issue with the boundary conditions can be easily resolved if we make the entire boundary essential. 

Mark and I identified the problem, and it will require to accommodate non homogeneous natural BCs, and that in turn will need a couple of integrators that we don't currently have. 

You think I should push the quick fix so that it can go to the release? 

Thanks, 

Socratis

**Comment by tzanio** [comment]
> The issue with the boundary conditions can be easily resolved if we make the entire boundary essential... it will require to accommodate non homogeneous natural BCs, and that in turn will need a couple of integrators that we don't currently have.

Hmmm, how did it work before? 

> You think I should push the quick fix so that it can go to the release?

As long as it is not a lot of code and it will not take a lot of time, yes :)

**Comment by psocratis** [comment]
> > The issue with the boundary conditions can be easily resolved if we make the entire boundary essential... it will require to accommodate non homogeneous natural BCs, and that in turn will need a couple of integrators that we don't currently have.
> 
> Hmmm, how did it work before?
> 
> > You think I should push the quick fix so that it can go to the release?
> 
> As long as it is not a lot of code and it will not take a lot of time, yes :)

Well it was wrong before. 

The code changes are minimal. I can do the changes in this PR shortly after the meeting. 

Socratis

**Comment by psocratis** [comment]
> > > The issue with the boundary conditions can be easily resolved if we make the entire boundary essential... it will require to accommodate non homogeneous natural BCs, and that in turn will need a couple of integrators that we don't currently have.
> > 
> > 
> > Hmmm, how did it work before?
> > > You think I should push the quick fix so that it can go to the release?
> > 
> > 
> > As long as it is not a lot of code and it will not take a lot of time, yes :)
> 
> Well it was wrong before.
> 
> The code changes are minimal. I can do the changes in this PR shortly after the meeting.
> 
> Socratis

I pushed  the changes. I think it looks good now. 

Socratis

File: examples/ex22p.cpp [hunk 1]
```diff
@@ -437,6 +450,10 @@ int main(int argc, char *argv[])
       fgmres.Mult(B, U);
    }
 
+
+
+
+
```
Comment (v-dobrev) [inline examples/ex22p.cpp]: We should probably revert this added extra space.

**Comment by v-dobrev** [comment]
I did not test the new code but it looks good to me.

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Comment by tzanio** [comment]
Are the results from tonight's regression runs OK with this?

**Comment by psocratis** [comment]
> Are the results from tonight's regression runs OK with this?

Everything looks good to me! 

Iteration numbers and reported errors are not supposed to be the same with the baseline, since by fixing the BCs issue the problems are essentially different now. 

Other than that the reported iteration counts and errors seem reasonable to me. 

Socratis

**Comment by v-dobrev** [comment]
Should be good to go.

**Pull request merged**